src/diffusers/schedulers/scheduling_consistency_models.py

```diff
@@ -0,0 +1,330 @@
+        clip_denoised: bool = True,
+        timesteps: Optional[Union[List, np.ndarray, torch.Tensor]] = None,
+    ):
+        # standard deviation of the initial noise distribution
+        self.init_noise_sigma = sigma_max
+
+        # setable values
+        self.num_inference_steps = None
+        self.timesteps = timesteps
+        self.is_scale_input_called = False
+
+    def index_for_timestep(self, timestep, schedule_timesteps=None):
+        if schedule_timesteps is None:
+            schedule_timesteps = self.timesteps
+
+        indices = (schedule_timesteps == timestep).nonzero()
+        return indices.item()
+
+    def scale_model_input(
+        self, sample: torch.FloatTensor, timestep: Union[float, torch.FloatTensor]
+    ) -> torch.FloatTensor:
+        """
+        Scales the consistency model input by `(sigma**2 + sigma_data**2) ** 0.5`, following the EDM model.
+
+        Args:
+            sample (`torch.FloatTensor`): input sample
+            timestep (`float` or `torch.FloatTensor`): the current timestep in the diffusion chain
+        Returns:
+            `torch.FloatTensor`: scaled input sample
+        """
+        # Get sigma corresponding to timestep
+        if isinstance(timestep, torch.Tensor):
+            timestep = timestep.to(self.timesteps.device)
+        step_idx = self.index_for_timestep(timestep)
+        sigma = self.sigmas[step_idx]
+
+        sample = sample / ((sigma**2 + self.config.sigma_data**2) ** 0.5)
+
+        self.is_scale_input_called = True
+        return sample
+
+    def scale_timestep(self, timestep: Union[float, torch.FloatTensor]):
+        """
+        Scales the timestep based on the associated Karras sigma, for input to the consistency model.
+
+        Args:
+            timestep (`float` or `torch.FloatTensor`): the current timestep in the diffusion chain
+        Returns:
+            `torch.FloatTensor`: scaled input timestep
+        """
+        # Get sigma corresponding to timestep
+        if isinstance(timestep, torch.Tensor):
+            timestep = timestep.to(self.timesteps.device)
+        step_idx = self.index_for_timestep(timestep)
+        sigma = self.sigmas[step_idx]
+
+        return 1000 * 0.25 * torch.log(sigma + 1e-44)
+
+    def set_timesteps(self, num_inference_steps: int, device: Union[str, torch.device] = None):
+        """
+        Sets the timesteps used for the diffusion chain. Supporting function to be run before inference.
+
+        Args:
+            num_inference_steps (`int`):
+                the number of diffusion steps used when generating samples with a pre-trained model.
+            device (`str` or `torch.device`, optional):
+                the device to which the timesteps should be moved to. If `None`, the timesteps are not moved.
+        """
+        self.num_inference_steps = num_inference_steps
+
+        # Note: timesteps are expected to be increasing rather than decreasing, following original implementation
```

Comment:
Could you elaborate here why they need to be increasing? Usually the ordering is just an implementation detail and its ok for the diffusers convention of returning decreasing ordered timesteps

Reply:
When explicitly supplied, `timesteps` ultimately needs to be increasing because it is transformed into the `ramp` used for calculating the Karras sigma schedule in `set_timesteps`:

https://github.com/huggingface/diffusers/blob/dc4349cadc35d4515ef76cdf5981fc5b2fac5d99/src/diffusers/schedulers/scheduling_consistency_models.py#L187-L189

This follows the original implementation in [`stochastic_iterative_sampler`](https://github.com/openai/consistency_models/blob/main/cm/karras_diffusion.py#L657).

However, we could easily accept `timesteps` as decreasing in `__init__` and reverse it when we need to calculate the `ramp`.

Reply:
Changed `timesteps` to be expected to be decreasing, following the examples in https://github.com/huggingface/diffusers/pull/3492#issuecomment-1584880384.